lib/Protobuf.cpp

```diff
@@ -0,0 +1,422 @@
+      auto maybe_param = DecodeParameter(param);
+      if (maybe_param.Succeeded()) {
+        decl.params.emplace_back(maybe_param.TakeValue());
+      } else {
+        auto err = maybe_param.TakeError();
+        std::stringstream ss;
+        ss << "Could not parse " << i << "th parameter of function at address "
+           << address_str << ": " << err;
+        return {ss.str()};
+      }
+      ++i;
+    }
+
+    // Get the return address location.
+    if (function.has_return_address()) {
+      auto ret_addr = function.return_address();
+      auto maybe_ret =
+          DecodeValue(ret_addr, "return address", true /* allow_void */);
+      if (!maybe_ret.Succeeded()) {
+        auto err = maybe_ret.TakeError();
+        std::stringstream ss;
+        ss << "Could not parse return address of function at address "
+           << address_str << ": " << err;
+        return {ss.str()};
+
+      } else {
+        decl.return_address = maybe_ret.TakeValue();
+
+        // Looks like a void return type, i.e. function with no return address,
+        // so make sure there's no location/value info.
+        if (decl.return_address.type == void_type ||
+            decl.return_address.type == dict_void_type) {
+          if (decl.return_address.mem_offset || decl.return_address.mem_reg ||
+              decl.return_address.reg) {
+            std::stringstream ss;
+            ss << "Return address of function at address " << address_str
+               << " is marked as having a void type, but a location was "
+               << "specified";
+            return {ss.str()};
+          }
+
+          decl.return_address.type = void_type;
+
+          // Make sure the return address is address-sized.
+        } else {
+          auto dl = arch->DataLayout();
+          if (auto num_bits =
+                  dl.getTypeAllocSizeInBits(decl.return_address.type);
+              num_bits != arch->address_size) {
+            std::stringstream ss;
+            ss << "Return address of function at address " << address_str
+               << std::dec << " is a " << num_bits
+               << "-bit value, but address size for "
+               << remill::GetArchName(arch->arch_name) << " is "
+               << arch->address_size;
+            return {ss.str()};
+          }
+        }
+      }
+
+      // A lack of a return address suggests that this function has no return
+      // address, e.g. is something like `_start` on Linux, where there is no
+      // logical place to return, and no return address is initialized in the
+      // appropriate place in a register / on the stack by the kernel.
+    } else {
+      decl.return_address.type = void_type;
+    }
+
+    /*
+    // Decode the value of the stack pointer on exit from the function, which is
+    // defined in terms of `reg + offset` for a value of a register `reg`
+    // on entry to the function.
+    if (auto ret_sp = obj->getObject("return_stack_pointer")) {
```

Comment:
Does this need to be part of the spec format?

Reply:
@2over12